### PR TITLE
SONARJS-768 Rule: The output of functions that don't return anything should not be used

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UseOfEmptyReturnValueCheck.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+    if (parent.is(CONDITIONAL_OR, CONDITIONAL_AND)) {
+      return ((BinaryExpressionTree) parent).leftOperand().equals(tree);
+    }
+
+    if (parent.is(CONDITIONAL_EXPRESSION)) {
+      return ((ConditionalExpressionTree) parent).condition().equals(tree);
+    }
+
+    return !parent.is(Kind.EXPRESSION_STATEMENT);
```
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UseOfEmptyReturnValueCheck.java]: @vilchik-elena what if parent is a parenthesised expression?

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UseOfEmptyReturnValueCheck.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  }
+
+  @CheckForNull
+  private static FunctionTree functionTree(CallExpressionTree tree, ProgramState currentState) {
+    if (tree.callee().is(Kind.IDENTIFIER_REFERENCE)) {
+      Symbol symbol = ((IdentifierTree) tree.callee()).symbol();
+      
+      if (symbol != null) {
+        SymbolicValue symbolicValue = currentState.getSymbolicValue(symbol);
```
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UseOfEmptyReturnValueCheck.java]: @vilchik-elena Why not use the expression stack?

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UseOfEmptyReturnValueCheck.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+      }
+    }
+
+    return null;
+  }
+
+  @Override
+  void raiseIssue(CallExpressionTree tree) {
+    String functionName = ((IdentifierTree) tree.callee()).name();
+    addIssue(tree.callee(), String.format(MESSAGE, functionName, functionName));
```
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UseOfEmptyReturnValueCheck.java]: @vilchik-elena Maybe we could add a secondary location on the function declaration (or a list of secondary locations as there may be more several function declarations depending on the execution path...).

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UseOfEmptyReturnValueCheck.java (hunk 1 above)
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UseOfEmptyReturnValueCheck.java]: @pynicolas thanks! fixed

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UseOfEmptyReturnValueCheck.java (hunk 2 above)
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UseOfEmptyReturnValueCheck.java]: @pynicolas fixed

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UseOfEmptyReturnValueCheck.java (hunk 3 above)
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UseOfEmptyReturnValueCheck.java]: @pynicolas This is not easy to implement. We can keep the list of `FunctionTree`, which might be declaration or expression or arrow function. But it's not clear which part to highlight.

**Pull request merged**